### DISCUSSION TITLE
[Bug]: Integration continuously failing after SW update

### DISCUSSION THREAD

**Discussion by 9-ck** [wlcrs/huawei_solar#516] [Connecting to the inverter]
**###** Describe the issue

I've had the SW of my SUN2000 10KTL-M1, SUN2000 3KTL-L1 and the SDongle BT22… updated a couple of days ago. I've experienced issues with the integration since. It fails to initialize / setup. I've noticed that it is "lagging" during restart.

I thought that it had something to do with my configuration.yaml file since when I deleted newly added parts and rebooted HA the integration would setup. But working in the configuration.yaml (without the parts that I previously thought was the cause), the integration now won't initialize after a simple restart.

Prior to this instance I've not had a single issue with either HA nor the integration.

The Huawei SW updates were as follows:
SDongle: from **V1**00R0**01C0**0SPC**133** (December 2022) to **V2**00R0**22C1**0SPC**108** (September 2023)
SUN2000 10 KTL-M1: from V100R001C00SPC**153** (March 2023) to V100R001C00SPC**156** (October 2023)
SUN2000 3 KTL-L1: from V200R001C00SPC**130** (March 2023) to V200R001C00SPC**132** (October 2023)

I've contacted my installer in order to get him to confirm that the SW update of the SDongle infact is correct, as the the "V200…" is a different version. I use a SDongle connected via cable. Also the log mention "no modbus connection". Installer has confirmed modbus is activated and non-restricted after SW update.

I will upload the diagnostics file when / if i get the integration up and running. I could reinstal the integration if that would make any sence? I have the Debug log from the integration, but as I understant it you wan't the full log.

![2023-10-21_12-11-54](https://github.com/wlcrs/huawei_solar/assets/148528096/8195d4ce-485f-47a0-9736-f41e63184e89)

### Bescribe your Huawei Solar Setup

Inverter Type: SUN2000 10KTL-M1 / SUN2000 3KTL-L1 
Inverter Firmware version: V100R001C00SPC156 / V200R001C00SPC132
SDongle present: yes (Firmware version: V200R001C00SPC132)
Power meter present: three phase
Battery: LUNA2000 10 kWh
Battery Firmware version: V100R002C00SPC119

### How do you connect to the inverter?

Please select your connection method
Don't know why my answer isn't here. I connect via network cable.

### Upload your Diagnostics File

Drag & Drop your Diagnostics File here.

**Comment by 9-ck** [comment]
Here is the Full Log
[config_entry-huawei_solar-1a6c1dee2434020362aa55e51f2e97c8.json.txt](https://github.com/wlcrs/huawei_solar/files/13060546/config_entry-huawei_solar-1a6c1dee2434020362aa55e51f2e97c8.json.txt)

**Comment by 9-ck** [comment]
Here is the Huawei Solar debug log
[home-assistant_huawei_solar_2023-10-21T09-43-36.482Z.log](https://github.com/wlcrs/huawei_solar/files/13060548/home-assistant_huawei_solar_2023-10-21T09-43-36.482Z.log)

**Comment by SirInsky12** [comment]
I have exactly the same problem, but found the issue. Clear your Modbus entries from config.yaml.

Kill all count with a # before.

modbus:
  - name: pluggit
    type: tcp
    host: 10.10.0.35
    port: 502
    sensors:
      - name: Pluggit_Outdoor_T1
        slave: 1
        address: 133
        unit_of_measurement: °C
        count: 2
        data_type: float32

Reboot it and it  works

**Comment by Wintherman** [comment]
I have the same problem.
But where can I find config.yaml?
I have searched in the huawie_solar directory but cannot find.

**Comment by 9-ck** [comment]
Thanks @SirInsky12! I'm sorry, but I'm not sure which config.yaml you're referring too. Is it th configuration.yaml? I have no modbus entries here. I installed Huawei Solar via HACS / not manually. Or is it maybe the config_flow.py file inside the huawei_solar folder? That’s some file to go through.

The funny thing is that the integration has been stable since my post. I re-edited some sensors with what I believe is "older" code (not to solve the issue, just to clean up). Deleted the integration, rebooted, installed Huawei Solar again. Also moved all my templates to a templates.yaml file.

Here is a small piece of the code I copied…
          {% set price = float(states('sensor.energi_data_service'),0) %}
          {% set meter = float(states('sensor.energy_battery_charge_grid'),0) %}
          {% set meter = float(states('sensor.energy_battery_charge_grid'),0) %}

I changed it to this…
          {% set price = states('sensor.energi_data_service')|float(0) %}
          {% set meter = states('sensor.energy_battery_charge_grid')|float(0) %}
          {% set delta = meter - this.attributes.last|float(0) %}

I am no wiz, more like a novice, so I'm not sure it´s "old" code and I'm just telling you what happened. Hard for me to say if it has any relevance at all.

**Comment by 9-ck** [comment]
OK - the Huawei Solar integration crashed this morning while I was editing my templates.yaml file. Attached is the debug log.

[home-assistant_huawei_solar_2023-10-22T07-54-30.767Z.log](https://github.com/wlcrs/huawei_solar/files/13062674/home-assistant_huawei_solar_2023-10-22T07-54-30.767Z.log)

**Comment by 9-ck** [comment]
Reading the debug log, trying to understand and correct / clean. Rebooted and installed the integration again. I now see that the naming of the inverter input sensors are back to what they were before I had issues…? When I first installed the integration the name was "sensor.inverter_input_power_1" - then it changed to "sensor.inverter_1_input_power_1" and now it's back to "sensor.inverter_input_power_1".

UPDATE: Crash again. I suspect that I have some mistakes in my templates.yaml file which provoke the Huawei Solar integration.

**Comment by Wintherman** [comment]
Hi,
I am up and running again after Delete and installed to the latest version 1.2.7 I think and reboot.
I could run my little script forcing the battery to charge from the grid up to 70% with "huawei_solar.forcible_charge_soc"
But still it would be nice to know where config.yaml is placed.

**Comment by SirInsky12** [comment]
Please check your configuration.yaml in Homeassistant and edit your Modbus entries, is not allowed to use count for that.

**Comment by 9-ck** [comment]
> Please check your configuration.yaml in Homeassistant and edit your Modbus entries, is not allowed to use count for that.

Thanks @SirInsky12 but I don't have any modbus entries in my configuration.yaml file? If you check the lates debug log I uploaded in the thread you'll notice that (at least to my eyes) the Huawei Solar integration has some issues with bad formatting in my templates - I use DKK and kWh/kWp. I did clean this up, but just had another crash.

**Comment by 9-ck** [comment]
Hmm... My CPU load on VM for HA is around 85-90% constantly and 5% user on top of that. That seems a bit to squeezed - no?
I'm running HA-core (latest ver. 2023.10.5) in an Oracle VM VirtualBox Manager on an Intel NUCi5 with 16 GB ram and 8 processors. The VM for HA has 4GB ram and 2 processors allocated. The only other thing running on the NUC is PLEX, also in a VM (4 GB ram and 4 processors allocated to PLEX).

My HA iso image is approximately 30 GB. Would you consider the CPU load normal?

Admittedly I experienced the latest crash while working in HA, playing with views several views.

**Comment by SheikhMishu** [comment]
Same issue, I have added some logs if that helps. 

![image](https://github.com/wlcrs/huawei_solar/assets/21236749/dd4d9869-9ef9-446c-b615-a8353c286f1a)


[huawei_solar.huawei_solar_1.log](https://github.com/wlcrs/huawei_solar/files/13064606/huawei_solar.huawei_solar_1.log)
[huawei_solar.huawei_solar_2.log](https://github.com/wlcrs/huawei_solar/files/13064607/huawei_solar.huawei_solar_2.log)

**Comment by SirInsky12** [comment]
Go to developer tools and check the config and look what happens

**Comment by 9-ck** [comment]
> Same issue, I have added some logs if that helps.

@SheikhMishu As said I'm no wiz, but it seems you have time out errors. Do you have a good enough LAN / WiFi connection?

**Comment by 9-ck** [comment]
@SirInsky12 Appreciate you trying to help but where do you mean? In Developer Tools I have YAML / STATES / SERVICES / TEMPLATE/ EVENTS / STATISTICS and ASSIST.

In the STATISTICS I have a bunch of issues with regards to "units" some that I have created myself, some which come from other integrations. Seems strange that HA wouldn't know kWh, m and °C - also why doesn't HA know the Danish currency DKK? I'm unsure what will happen if I press "Fix Issue".

![2023-10-23_14-05-37](https://github.com/wlcrs/huawei_solar/assets/148528096/0710299c-3ba0-430f-bda7-fc8b6a6e0b10)

BTW, in my VM setup I have increased the numbers of processors to 4 and ram to 6GB. CPU load is down to 60-70%. But still I am experiencing dropouts now and then. HA will run fine for a while and suddenly drop out.

**Comment by 9-ck** [comment]
Could someone please have a look at this and comment. Is it OK with the "errors"?
![2023-10-25_10-57-31](https://github.com/wlcrs/huawei_solar/assets/148528096/48aa2c8f-6699-4c11-8bfc-5faa634dbf69)

**Comment by matt7py** [comment]
The same problem was working normally, but suddenly, it would not accept connections. I attempted to resolve it by deleting the integration and reinstalling it. Additionally, I tried reconnecting via WiFi, but the issue persists. !!

![image](https://github.com/wlcrs/huawei_solar/assets/120527800/371d607a-d445-4578-8c0d-f79edb58d42a)

**Comment by 9-ck** [comment]
This issue continues. It seams like frequent restarts in connection with working in the configuration file will provoke a crash in my case.

**Comment by 9-ck** [comment]
I have now tried to install the integration on a completely new Home Assistant and the integration crashed on me. Could someone please enlighten me on as to why this is happening? Attached the full logs from the new HA. @wlcrs has the thread been moved to "discussions" because it's not an issue with the integration?

![2023-10-30_10-46-27](https://github.com/wlcrs/huawei_solar/assets/148528096/d90cf527-8864-43dd-9ecb-35020ac94243)

[home-assistant_2023-10-30T09-46-55.475Z.log](https://github.com/wlcrs/huawei_solar/files/13203781/home-assistant_2023-10-30T09-46-55.475Z.log)

**Comment by WildEchoWanderer** [comment]
My integration was also working with no problems. Now it has the same Error for around 5 days (last value on 27. October 2023 @06:22:54).

I can fix it with a Homeassistant reboot, but it works only for around 1h & 40min and stops again there.

Hit me up if i can help to solve the issue

I am Connected to the Inverter WIFI directly with a USB-WIFI Stick in my HAOS Ienovo. It worked for month without problems

![Screenshot_20231031_101824_io homeassistant companion android](https://github.com/wlcrs/huawei_solar/assets/116340711/10a5abd5-2959-4c85-931e-ae3e578c1342)

EDIT: Had to delete log because of exposed keys

**Comment by 9-ck** [comment]
OK, back to report that I believe I have solved my issue. I have been configuring a lot of sensors with reference to the Huawei Solar integration lately. It seams that the integration is very sensitive to errors in sensors that depend on input from the integration. Going thru the logs I found a few errors. Corrected them and now the integration has been stable since 12 o'clock.

**Reply by WildEchoWanderer** [reply]
I deleted nearly every sensor related to the Huawei Solar integration and it still stopped working. 
Yours still running fine?

**Reply by 9-ck** [reply]
Nope. It has become more stable, but it crashes often when I do a restart, but not on a quick reload or reboot.

Edit: Just updated to core 2023.11.1 and restart went fine (fingers crossed).

**Reply by WildEchoWanderer** [reply]
Funny to mention: i updated to 2023.11.1 too and after a reboot it is working stable. Will Leave it Like this for a Week and the Start adding the template Sensors again

**Comment by WildEchoWanderer** [comment]
I have this error after reboot (but it works for aprox. 1 hour like this):
![Screenshot_20231103_141359](https://github.com/wlcrs/huawei_solar/assets/116340711/a53531c4-0374-441f-a337-e1af56735ab2)

**Comment by WildEchoWanderer** [comment]
@wlcrs For me the error seems fixed after Updating to 2023.11.1
i don't know how and why

**Reply by 9-ck** [reply]
@Andurilll - Funny... I went in here to report back just to find out that the post is #1 - Way to go! 😊. I too have experienced a more stable integration, but it still crashes occasionally. It has just happened to me twice this evening - after restart. 

It seems to me that the integration (or the connection) is sensitive to frequent restarts one after another. Quick reloads don't cause any issue.

As mentioned above I have studied the debug logs carefully and it also seems to me that the integration is somewhat sensitive to how sensors are used. I get error messages using `| float` and not `| float(0)`. In mathematical divisions using two sensors from the integration I have now decided to try out to use `| float(0)` for the numerator and `| float (1)` for the denominator - so the result will not be false (because of dividing with 0). Don't know if that will have any effect at all.

Anyway, thank you a lot for your work @wlcrs and I hope our feedback is useful somehow.

**Reply by WildEchoWanderer** [reply]
@9-ck 
Sadly it happened again after activating these 3 calculated sensors in my config.yaml :

```
sensor:
#  - platform: template
#    sensors:
#      energy_harvested_today:
#        friendly_name: 'Harvested Energy Today'
#        value_template: "{{ (states('sensor.inverter_daily_yield')|float + states('sensor.battery_day_charge')|float - states('sensor.battery_day_discharge')|float)|round(3) }}"
#        unit_of_measurement: "kWh"

template:
#  - sensor:
#    - name: "Battery Charge Only"
#      unique_id: "battery_charge_only"
#      unit_of_measurement: "W"
#      device_class: power
#      state_class: measurement
#      state: >-
#        {{ max(0, states('sensor.battery_charge_discharge_power')|float(0)) }}
#  - sensor:
#    - name: "Battery Discharge Only"
#      unique_id: "battery_discharge_only"
#      unit_of_measurement: "W"
#      device_class: power
#      state_class: measurement
#      state: >-
#        {% set batterypower = states('sensor.battery_charge_discharge_power') | float %}
#        {% if batterypower < (0) %}
#          {{ batterypower * -1 }}
#        {% else %}
#          {{ 0 }}
#        {% endif %}

```
How would i change it to work with your `| float` ?

Maybe this happens, when a sensor calculates with other sensors of the integration before it is loaded. Maybe this article helps: https://community.home-assistant.io/t/float-0-and-automations-in-2021-12/357306/6

**Reply by Roving-Ronin** [reply]
@Andurilll your formating is using the old emthod of platform: template and then moves to the new/current format for the next two sensors. Also mixes name and friendly_name. 

In the current format it could all be written like this below, with device class and state class also added to the Harvested Sensor.
Please give these updates sensors a try and see if it works, and when you get it working please add into the Wiki section for everyone to benefit from: https://github.com/wlcrs/huawei_solar/wiki/Daily-Solar-Yield#energy-calculation-from-the-inverter-and-battery-sensors-values-without-riemann-integration

```

template:

   # Sensor that calculates and net harvested energy for the day in kilowatt-hours. 
   # The result is the sum of the inverter daily yield and the battery daily charge, minus the battery daily discharge
  - sensor:
    - name: "Harvested Energy Today"
      unique_id: energy_harvested_today
      unit_of_measurement: "kWh"
      device_class: energy
      state_class: total_increasing
      state: {{ ( (states('sensor.inverter_daily_yield') | float ) + (states('sensor.battery_day_charge') | float ) ) - (states('sensor.battery_day_discharge') | float) | round(3) }}


   # Sensor that reports the battery charge discharge power
  - sensor:
    - name: "Battery Charge Only"
      unique_id: "battery_charge_only"
      unit_of_measurement: "W"
      device_class: power
      state_class: measurement
      state: {{ max(0, states('sensor.battery_charge_discharge_power') | float(0)) }}

   # Sensor that reports if the battery is discharging
  - sensor:
    - name: "Battery Discharge Only"
      unique_id: "battery_discharge_only"
      unit_of_measurement: "W"
      device_class: power
      state_class: measurement
      state: >-
        {% set batterypower = states('sensor.battery_charge_discharge_power') | float %}
        {% if batterypower < (0) %}
          {{ batterypower * -1 }}
        {% else %}
          {{ 0 }}
        {% endif %}

```

**Reply by JensenNick** [reply]
@Roving-Ronin Could you please explain how I could ad my project to the wiki. I believe both of you will find it interesting. Is it OK "just" to ad a new page? No QC :-).  GitHub Novice.

**Reply by Roving-Ronin** [reply]
@JensenNick in the wiki anyone with a GitHub account (as you have already) can edit and wiki pages and/or add a new page to the wiki.

Look on the wiki pages and d you'll see an edit and add page button.

If you want to send me the link (or paste it here) to the edits your making or new page, and I'll try peer review it.

**Comment by 9-ck** [comment]
> @9-ck Sadly it happened again after activating these 3 calculated sensors in my config.yaml 

@Andurilll Yes, as mentioned above the integration seams a bit sensitive to the formatting and especially if you need to use `| float` without `(0)`. I got around this by using the an availability sensor. Admittedly I did find an error in my firewall and this may have been my issue all along. The integration has now run stable for several days with lot of work in HA and continuous restarts. Restart upon restart it takes a while to reconnect but it does not fail.

That said I see that we are / have been working on the same type of project. Take a look below.

**Reply by Roving-Ronin** [reply]
@9-ck you mention the configuration.yaml. I would strongly suggest you look into making a /packages directory in the HA root folder and adding 

'''
packages: !include_dir_named packages
'''

To your configuration.yaml

- [ ] With this you can then start making seperate yaml files for the different things you add / mod in HA. ie for this integration add a huawei_derived_sensors.yaml and put all these sensors in it. Same again if you define power costs et, create a huawei_electricity_costs.yaml etc

Makes it a lot easier to understand the yaml versus all in one configuration.yaml also means you can rename the file extension as a cheat to enable/disable the contents and error check / see if HA starts or not.

**Reply by 9-ck** [reply]
@Roving-Ronin Thanks! I do have my configuration split up, although not in packages only files - but what you explain makes sence. I did try to split the automations.yaml in order to keep my "# text" - didn't succeed :-) but I'll look at it again. Thanks again!

**Reply by Roving-Ronin** [reply]
@9-ck have a look at this thread, it's got all my yaml files for the derived sensors and costs sensors etc that I use with WLCRS.

https://github.com/wlcrs/huawei_solar/discussions/518

If you look at them you'll see it means all the cost for this is in discrete files and things like automations can also be defined there..

Look for the section with automations:

**Comment by JensenNick** [comment]
> If you want to send me the link (or paste it here) to the edits your making or new page, and I'll try peer review it.

@Roving-Ronin Thanks. Done my best. Would apricate to hear from you if it somehow isn't as it should be. For another time - How do I contact you / others with a "private" message? Damn I'm getting old...

https://github.com/wlcrs/huawei_solar/wiki/Huawei-Solar-%E2%80%90-PEES-(Power,-Energy-and-Economy-Sensors)

**Reply by Roving-Ronin** [reply]
Will look when I have time... busy this coming week though.

Re PM.. let me know when you work it out, I've been wondering the same thing myself.

Seems like there was a message feature but it's been removed from GitHub, now there's some mention of setting up a email in GitHub that relays to your actual/private email.

**Reply by JensenNick** [reply]
@Roving-Ronin I've converted my sensors into a package as well. Is this how you would go about providing this kind of input to others? PS: I haven't had time to test it. https://github.com/JensenNick/huawei_solar_pees

**Reply by Roving-Ronin** [reply]
@JensenNick just had a quick glance and looks like your utility meter yaml files will fail, as they don't start by defining them as 

```
utility_meter:
```
If you want to use the utility meter to actually track amounts used/exported and the amount paid/compensated for it, I'd suggest looking at https://github.com/zeronounours/HA-custom-component-energy-meter that expands upon the basic utility meter to also add calculating costs and it will create an additional sensor that shows the $ side.  When using the 'cycle' option (hourly, daily, monthly, yearly etc) you'll get a sensor tracking hourly-yearly use and costs/compensation and if you also use the 'tariffs' option in the sensor you can have it lookup a sensor that provides the current price/compensation and use that for recording the $ side.

**Reply by JensenNick** [reply]
@Roving-Ronin Thanks! OK - ”language barrier”. I 'm not sure I understand what you mean by 

> as they don't start by defining them as utility_meter:

I believe I have to followed the documentation? https://www.home-assistant.io/integrations/utility_meter/ 

![2023-11-20_12-17-17](https://github.com/wlcrs/huawei_solar/assets/149193835/1efac5e1-7e56-40f6-aad8-230c8ab93436)

Not to say that you suggestion isn’t better/smarter, but I have the utility meters set up and running in my current setup and without issues so far. Admittedly I was really struggling with the utility meters based on this sensor 

```
name: "Energy Yield Total"
unique_id: energy_yield_total
unit_of_measurement: kWh
device_class: energy
state_class: total_increasing
state: >
  {{ states('sensor.energy_yield_1') | float + 
  states('sensor.energy_yield_2') | float }}
availability: >
  {{ has_value('sensor.energy_yield_1') and 
  has_value('sensor.energy_yield_2') }}
```

The sensor is the sum of two energy sensors and every time the input sensors lost their value the sensor would default to zero or “unavailable”. The above sensor was jumping “back on track” once it was up running again but the utility meters based on it would go crazy. I finally solved the issue with the addition of the “availability” which lets me keep `| float`. I have tested it and I haven’t had any issues since.

I will have a closer look at your suggestion. I see that it incorporates peak hours etc. I guess we are a bit spoiled in DK because of the great work that MTrab has done with the Energi Data Service which is capable of returning the current electricity price incl. the different tariffs , peak and off-peak, and incl. VAT. 

https://github.com/MTrab/energidataservice

**Comment by WildEchoWanderer** [comment]
Funny speaking i wanted to report how smooth everything was with all the Template sensors, but after the last HA Update it failed again...

**Comment by WildEchoWanderer** [comment]
To anybody Setting Here searching For a Solution: after upgrading my inverter, battery & dongle it seems very stable

It could be caused by 2 Different Software versions in the 2 Batteries itself
